### PR TITLE
GS: Threaded+async video capture/audio capture/custom resolution

### LINKED ISSUE #3663: [Feature Request] Capture video and audio in 1 combined file.
The current implementation:
User activates Video Capture.
GSdx will spit out a video file(extensionless for now, but there is a PR that will add .avi to it)
SPU2-X will spit out a audio file (recorder.wav will be changed to audio_recorder.wav also in same PR)

Future implementation:
Perhaps easiest way is to wait till the plugin system is merged with the core.

### LINKED ISSUE #4032: [Feature Request] Overhaul of the screenrecorder
**Description**
This is not the same as #3663 but more an extension upon it. Making it easier to decide what codec, size to pick for the enduser intended (or as close as possible) result.

Current look:
image

Not very intuitive in my opinion but does a good job for what it needs to do.

**Reason**
Lowers flooding storage because you can see if it's still recording and as reminder it's recording (especially for uncompressed), maybe set a way to stop recording after a certain time ( a minute max recording) or certain filesize (2 GB). Display in OSD of some kind to display how many seconds/minutes have passed while recording and append the current usage in increments every second, at the end give an overview of the format of video, duration, filesize and the filepath.

**Examples**
OBS, Audacity, internal W10 recorder. Lots of different styles that do show the duration but not the filesizes.

### PR DESCRIPTION
**Pull request by stenzek**
### Description of Changes

Enhances video capture by adding:
 - Asynchronous readback of frames (currently +3 frames behind presentation), so it doesn't block your GPU.
 - Threaded encoding. ffmpeg is already threaded internally for some codecs, but this moves the image scaling/colour space conversion off thread too.
 - Adds audio to video captures. You can capture video+audio (the default), or video or audio only.
 - Adds an option to manually set the video resolution. Can be useful when playing at very high upscales, but you only want to capture at e.g. 1080p.

Unfortunately this meant I had to pretty much rewrite how downloads happen in GS :( So Metal isn't done yet. But we can still test it on the other platforms.

Also needs ffmpeg v5.x compatibility added. Channel layout API was changed.

Old wav record stuff was removed, because it's redundant when we can use the new capture for audio only, as well as it supporting a much wider range of formats.

### Rationale behind Changes

Closes #3663.
Closes #4032.

### Suggested Testing Steps

Test games which rely on GS downloads, across all renderers, make sure they didn't break. I've only checked Oni (lights through walls), and it _seems_ okay.

Test new video dumping options :)

### CODE REVIEW

**Comment by Tokman5** [comment]
https://github.com/PCSX2/pcsx2/blob/1cfa054385e0c2f0b48e6cbe1eb61f9b22da6f06/pcsx2/GS/GSCapture.cpp#L586
the third parameter should be `&s_audio_codec_arguments`

File: pcsx2/Config.h [hunk 1]
```diff
@@ -605,15 +605,18 @@ struct Pcsx2Config
...
 
 		static constexpr float DEFAULT_FRAME_RATE_NTSC = 59.94f;
 		static constexpr float DEFAULT_FRAME_RATE_PAL = 50.00f;
 
-		static constexpr u32 DEFAULT_VIDEO_CAPTURE_BITRATE = 6000;
-		static const char* DEFAULT_VIDEO_CAPTURE_CONTAINER;
+		static constexpr int DEFAULT_VIDEO_CAPTURE_BITRATE = 6000;
+		static constexpr int DEFAULT_VIDEO_CAPTURE_WIDTH = 640;
+		static constexpr int DEFAULT_VIDEO_CAPTURE_HEIGHT = 480;
+		static constexpr int DEFAULT_AUDIO_CAPTURE_BITRATE = 64;
```
Comment (refractionpcsx2) [inline pcsx2/Config.h:618]: Can you change this to 160? 64 is pretty gross.

**Comment by refractionpcsx2** [comment]
Not sure if this is another Aspect Ratio bug or not, but on discord when you post the video it shows in 4:3 fine, but the moment you start playing, it goes in to 16:9 mode.

File: pcsx2/Config.h (hunk 1 above)
Comment (RedDevilus) [reply pcsx2/Config.h:618]: Even phones as media player do 192-384 khz for some years now. Beyond that is probably superears/audiophile or placebo.

Comment (stenzek) [reply pcsx2/Config.h:618]: Sure, will do after work.

**Comment by stenzek** [comment]
> Not sure if this is another Aspect Ratio bug or not, but on discord when you post the video it shows in 4:3 fine, but the moment you start playing, it goes in to 16:9 mode.

It's probably just Discord's thumbnail stuff not respecting the PAR/DAR of the video. It'll capture at whatever the internal resolution is by default, and then it's up to the player to aspect correct it. Alternatively, you can specify an explicit resolution - if that is the correct aspect, then I'm guessing the thumbnail will be too.

File: pcsx2/GS/GSCapture.cpp [hunk 2]
```diff
@@ -242,204 +346,531 @@ bool GSCapture::BeginCapture(float fps, GSVector2i recommendedResolution, float
...
 
-	// Default to YUV 4:2:0 if the codec doesn't specify a pixel format.
-	if (!codec->pix_fmts)
-	{
-		s_codec_context->pix_fmt = AV_PIX_FMT_YUV420P;
+		s_video_codec_context->codec_type = AVMEDIA_TYPE_VIDEO;
+		s_video_codec_context->bit_rate = GSConfig.VideoCaptureBitrate * 1000;
+		s_video_codec_context->width = s_size.x;
+		s_video_codec_context->height = s_size.y;
+		s_video_codec_context->sample_aspect_ratio = wrap_av_d2q(aspect, 100000);
```
Comment (Tokman5) [inline pcsx2/GS/GSCapture.cpp:395]: if i understand correctly, DAR value is wrongly set to SAR in this line

Comment (stenzek) [reply pcsx2/GS/GSCapture.cpp]: It does, if the aspect ratio of the TV is not the same as the aspect ratio of the framebuffer (which is fairly common).

**Review by Mrlinkwii — APPROVED** [review]

**Comment by RedDevilus** [comment]
Don't forget to extract this on root of PCSX2:
https://www.duckstation.org/upload/ffmpeg-dlls.zip

_*If things changed in the future then keep in mind the date of posting._

**Pull request merged**